### PR TITLE
[CARBONDATA-4338] moving dropped partition data to trash

### PR DESCRIPTION
**Pull request by maheshrajus**
### Why is this PR needed?
When drop partition operation is performed carbon data will modify only table status file and can not delete the actual partition  folder which contains data and index files. As comply with hive behaviour carbon data also should delete the deleted partition folder in storage[hdfs/obs/etc..]. Before deleting carbon data will keep copy in Trash folder. User  can restore it by checking the partition name and time stamp.
 ### What changes were proposed in this PR?
Moved the deleted partition folder files to trash folder
 ### Does this PR introduce any user interface change?
 - No
 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4590/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6335/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4594/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/725/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6337/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6341/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/729/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6342/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4599/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/730/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4605/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/736/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.3/4/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4623/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/754/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6368/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4624/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6370/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/757/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 1]
```diff
@@ -105,6 +109,28 @@ case class CarbonAlterTableDropHivePartitionCommand(
...
+          // move  the partition files to trash folder which are dropped
+          val droppedPartitionNames = partitions.map { partition =>
+            partition.spec.map { specs => specs._1 + "=" + specs._2}
+          }
+          val droppedPartitionName = droppedPartitionNames
+            .head
+            .headOption
+            .get
+          TrashUtil.copyFilesToTrash(carbonPartitionsTobeDropped,
+            TrashUtil.getCompleteTrashFolderPath(
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:122]: `getCompleteTrashFolderPath` is used to create Path for Non-partition table(appends Segment Prefix), I think, for partition table, we should create Path with Partition folder name

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 2]
```diff
@@ -105,6 +109,28 @@ case class CarbonAlterTableDropHivePartitionCommand(
             ifExists,
             purge,
             retainData).run(sparkSession)
+
+          // move  the partition files to trash folder which are dropped
+          val droppedPartitionNames = partitions.map { partition =>
+            partition.spec.map { specs => specs._1 + "=" + specs._2}
+          }
+          val droppedPartitionName = droppedPartitionNames
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:117]: Taking Head Option to create Partition folder name in Trash is wrong here. For example, consider a case, where user performs drop multi-level partititon (i.e., `DROP PARTITION(deptname='Learning',doj='2015-12-01 00:00:00' )`) in this case, Trash folder also should have multi-level partition folder, to roll back drop partition if user wants. So, in TRash, please copy under respective multi-level partition folders

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 3]
```diff
@@ -105,6 +109,28 @@ case class CarbonAlterTableDropHivePartitionCommand(
...
+            .head
+            .headOption
+            .get
+          TrashUtil.copyFilesToTrash(carbonPartitionsTobeDropped,
+            TrashUtil.getCompleteTrashFolderPath(
+              table.getTablePath,
+              CarbonUpdateUtil.readCurrentTime(),
+              droppedPartitionName))
+          // Delete partition folder after copy to trash
+          carbonPartitionsTobeDropped.asScala.foreach(delPartition => {
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:127]: Please format the new code changes

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 4]
```diff
@@ -30,13 +30,17 @@ import org.apache.spark.sql.parser.CarbonSparkSqlParserUtil
 import org.apache.spark.util.AlterTableUtil
 
 import org.apache.carbondata.common.logging.LogServiceFactory
+import org.apache.carbondata.core.constants.CarbonCommonConstants
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:33]: Please remove unused imports

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableDropTestCase.scala [hunk 5]
```diff
@@ -218,6 +222,129 @@ class StandardPartitionTableDropTestCase extends QueryTest with BeforeAndAfterAl
...
+    assert(network.length == 0)
+    val protocol = FileFactory.getCarbonFile(tablePath).listFiles().filter{
+      file => file.getName.equalsIgnoreCase("deptname=protocol")
+    }
+    assert(protocol.length == 0)
+    val security = FileFactory.getCarbonFile(tablePath).listFiles().filter{
+      file => file.getName.equalsIgnoreCase("deptname=protocol")
+    }
+    assert(security.length == 0)
+    sql("drop table if exists dropPartition1")
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableDropTestCase.scala:276]: please add a validation to check the data count after drop partition

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableQueryTestCase.scala [hunk 6]
```diff
@@ -494,9 +494,8 @@ test("Creation of partition table should fail if the colname in table schema and
       sql(s"""ALTER TABLE staticpartitionlocloadother_new DROP PARTITION(empname='ravi')""")
       checkAnswer(sql(s"select count(deptname) from staticpartitionlocloadother_new"), Seq(Row(10)))
       sql(s"""alter table staticpartitionlocloadother_new add partition (empname='ravi') location '$location'""")
-      checkAnswer(sql(s"select count(deptname) from staticpartitionlocloadother_new"), Seq(Row(20)))
       sql(s"""LOAD DATA local inpath '$resourcesPath/data.csv' INTO TABLE staticpartitionlocloadother_new partition(empname='ravi') OPTIONS('DELIMITER'= ',', 'QUOTECHAR'= '"')""")
-      checkAnswer(sql(s"select count(deptname) from staticpartitionlocloadother_new"), Seq(Row(30)))
+      checkAnswer(sql(s"select count(deptname) from staticpartitionlocloadother_new"), Seq(Row(20)))
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableQueryTestCase.scala:498]: why this change ?

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 4 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:33]: done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableQueryTestCase.scala (hunk 6 above)
Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableQueryTestCase.scala:498]: before we are not removing partition folders so when load again the count was 30.  With new change we are removing partition folders when we performed drop partition.  as per new change  modified the test case and test case is passing.

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableDropTestCase.scala (hunk 5 above)
Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/standardpartition/StandardPartitionTableDropTestCase.scala:276]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 1 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:122]: Added separate method for creating trash folder name with given partition folder names

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 2 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala]: Done and tested.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6377/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4632/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/763/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 7]
```diff
@@ -105,6 +106,25 @@ case class CarbonAlterTableDropHivePartitionCommand(
             ifExists,
             purge,
             retainData).run(sparkSession)
+
+          // move  the partition files to trash folder which are dropped
+          val droppedPartitionNames = partitions.map { partition =>
+            partition.spec.map { specs => specs._1 + "=" + specs._2}
+          }
+          val droppedPartitionName = droppedPartitionNames.head.mkString("/")
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:114]: i think, we should move the partition stale folders under respective trash partition folders. here, you are taking the first dropped partition folder to move other partitions data also.

Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:114]: Can we expose a property to enable/disable this operation during drop partition since it can degrade drop partition performance

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/767/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4636/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6381/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 7 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala]: done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4637/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6382/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/768/

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java [hunk 8]
```diff
@@ -268,4 +268,44 @@ public static String getCompleteTrashFolderPath(String tablePath, long timeStamp
...
+   *
+   * @param tablePath          absolute table path
+   * @param timeStampSubFolder the timestamp for the clean files operation
+   * @param partitionName      partition name for which files are moved to the trash folder
+   */
+  public static String getCompleteTrashFolderPathForPartition(String tablePath,
+      long timeStampSubFolder, String partitionName) {
+    return CarbonTablePath.getTrashFolderPath(tablePath) + CarbonCommonConstants.FILE_SEPARATOR
+        + timeStampSubFolder + CarbonCommonConstants.FILE_SEPARATOR
+        + CarbonTablePath.DATA_PART_PREFIX + partitionName;
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:283]: CarbonTablePath.DATA_PART_PREFIX is not needed to make partition path

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java [hunk 9]
```diff
@@ -268,4 +268,44 @@ public static String getCompleteTrashFolderPath(String tablePath, long timeStamp
...
+  }
+
+  /**
+   * The below method copies dropped partition files to the trash folder.
+   *
+   * @param filesToCopy              absolute path of the files to copy to the trash folder
+   * @param trashFolderWithTimestamp trashfolderpath with complete timestamp and segment number
+   */
+  public static void copyFilesToTrash(String filesToCopy, String trashFolderWithTimestamp)
+      throws IOException, InterruptedException {
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:293]: invalid method exception definition. Please remove

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java [hunk 10]
```diff
@@ -268,4 +268,44 @@ public static String getCompleteTrashFolderPath(String tablePath, long timeStamp
...
+        FileFactory.mkdirs(trashFolderWithTimestamp);
+      }
+      // check if file exists before copying
+      if (FileFactory.isFileExist(filesToCopy)) {
+        CarbonFile folder = FileFactory.getCarbonFile(filesToCopy);
+        CarbonFile[] dataFiles = folder.listFiles();
+        for (CarbonFile carbonFile : dataFiles) {
+          copyFileToTrashFolder(carbonFile.getAbsolutePath(), trashFolderWithTimestamp);
+        }
+      }
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:305]: can log the warning message if the statement `if (FileFactory.isFileExist(filesToCopy)) {` is false

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/774/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6391/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4646/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/777/

**Comment by Indhumathi27** [comment]
LGTM

File: core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java [hunk 11]
```diff
@@ -206,10 +206,12 @@ private static void getStaleSegmentFiles(CarbonTable carbonTable, List<String> s
   /**
    * This method will delete all the empty partition folders starting from the table path
    */
-  private static void deleteEmptyPartitionFoldersRecursively(CarbonFile tablePath) {
+  public static void deleteEmptyPartitionFoldersRecursively(CarbonFile tablePath) {
     CarbonFile[] listOfFiles = tablePath.listFiles();
     if (listOfFiles.length == 0) {
       tablePath.delete();
+      // if parent file folder also empty then delete that too.
+      deleteEmptyPartitionFoldersRecursively(tablePath.getParentFile());
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java:214]: this may delete the database also, right? why add this here?

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 12]
```diff
@@ -2886,5 +2886,14 @@ private CarbonCommonConstants() {
 
   public static final String CARBON_ENABLE_MULTI_VERSION_TABLE_STATUS_DEFAULT = "false";
 
+  /**
+   * This flag is set to enable the feature of moving the partition data to trash
+   * when dropped partition command is given.
+   * By default it is enabled if any performance degrade then user should disable this feature.
+   */
+  @CarbonProperty
+  public static final String CARBON_ENABLE_PARTITION_DATA_TRASH =
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:2895]: update in document file also, also add in FAQ, because here you are saying if impacted in performance, can disable.

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java [hunk 13]
```diff
@@ -268,4 +268,45 @@ public static String getCompleteTrashFolderPath(String tablePath, long timeStamp
...
+        + partitionName;
+  }
+
+  /**
+   * The below method copies dropped partition files to the trash folder.
+   *
+   * @param filesToCopy              absolute path of the files to copy to the trash folder
+   * @param trashFolderWithTimestamp trashfolderpath with complete timestamp and segment number
+   */
+  public static void copyFilesToTrash(String filesToCopy, String trashFolderWithTimestamp) {
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:292]: why are different methods required for handling partition tables? the existing method does not handle it? if so change the method name specific to the partition table, so that no one reuses it by mistake

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java [hunk 14]
```diff
@@ -268,4 +268,45 @@ public static String getCompleteTrashFolderPath(String tablePath, long timeStamp
...
+        CarbonFile[] dataFiles = folder.listFiles();
+        for (CarbonFile carbonFile : dataFiles) {
+          copyFileToTrashFolder(carbonFile.getAbsolutePath(), trashFolderWithTimestamp);
+        }
+      } else {
+        LOGGER.warn("Folder not copied to trash as partition folder does not exist");
+      }
+    } catch (IOException e) {
+      // If file is already moved or not found then continue with other files
+      LOGGER.warn("Unable to copy file to trash folder as file already moved or not found.", e);
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:309]: better to identify and give proper error messages rather than keeping users in ambiguous state

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 15]
```diff
@@ -106,6 +108,32 @@ case class CarbonAlterTableDropHivePartitionCommand(
...
             retainData).run(sparkSession)
+          val isPartitionDataTrashEnabled = CarbonProperties.getInstance()
+            .getProperty(CarbonCommonConstants.CARBON_ENABLE_PARTITION_DATA_TRASH,
+              CarbonCommonConstants.CARBON_ENABLE_PARTITION_DATA_TRASH_DEFAULT).toBoolean
+          if (isPartitionDataTrashEnabled) {
+            // move  the partition files to trash folder which are dropped
+            val droppedPartitionNames = partitions.map { partition =>
+              partition.spec.map { specs => specs._1 + "=" + specs._2 }
+            }
+            val timeStamp = CarbonUpdateUtil.readCurrentTime();
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:119]: please directly use System.CurrentMillis, no need to call method for one line

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 16]
```diff
@@ -106,6 +108,32 @@ case class CarbonAlterTableDropHivePartitionCommand(
             ifExists,
             purge,
             retainData).run(sparkSession)
+          val isPartitionDataTrashEnabled = CarbonProperties.getInstance()
+            .getProperty(CarbonCommonConstants.CARBON_ENABLE_PARTITION_DATA_TRASH,
+              CarbonCommonConstants.CARBON_ENABLE_PARTITION_DATA_TRASH_DEFAULT).toBoolean
+          if (isPartitionDataTrashEnabled) {
+            // move  the partition files to trash folder which are dropped
+            val droppedPartitionNames = partitions.map { partition =>
+              partition.spec.map { specs => specs._1 + "=" + specs._2 }
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:117]: do not hard code, please use org.apache.carbondata.core.constants.CarbonCommonConstants#EQUALS

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala [hunk 17]
```diff
@@ -106,6 +108,32 @@ case class CarbonAlterTableDropHivePartitionCommand(
...
+            }
+            val timeStamp = CarbonUpdateUtil.readCurrentTime();
+            for (index <- droppedPartitionNames.indices) {
```
Comment (akashrn5) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:120]: instead of for loop, if index required, you can use `zipwithindex` function

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 12 above)
Comment (maheshrajus) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:2895]: @akashrn5 can i disable this feature by default ?. if user want then enable and can use it. in this case we will mention only about moving partiton data to trash in documents.

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java (hunk 13 above)
Comment (maheshrajus) [reply core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:292]: yes, method name changed accordingly

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 15 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:119]: suppose if want to replace this System.CurrentMillis with some other in future then need to replace in all places. So i just used method call instead of directly using.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 16 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:117]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 17 above)
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala:120]: done

File: core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java (hunk 14 above)
Comment (maheshrajus) [reply core/src/main/java/org/apache/carbondata/core/util/TrashUtil.java:309]: done

File: core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java (hunk 11 above)
Comment (maheshrajus) [reply core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java:214]: this i added as part of bug fix. If partition folder is empty after moving to trash, it will delete empty folder. After moving if database folder also empty then it will delete that too.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4647/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/778/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala (hunk 15 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala]: for time and in this case it will be milliSec only and many places used directly, so use directly only

Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/partition/CarbonAlterTableDropHivePartitionCommand.scala]: ok

File: docs/ddl-of-carbondata.md [hunk 18]
```diff
@@ -1107,6 +1107,12 @@ Users can specify which columns to include and exclude for local dictionary gene
   ALTER TABLE locationTable DROP PARTITION (country = 'US');
   ```
 
+   The below flag is set to enable the feature of moving the deleted partition data to trash when dropped partition command is given.
+   ```
+   carbon.enable.partitiondata.trash
+   ```
```
Comment (akashrn5) [inline docs/ddl-of-carbondata.md:1113]: please add in configuration md file also

Comment (maheshrajus) [reply docs/ddl-of-carbondata.md:1113]: done

**Comment by akashrn5** [comment]
LGTM

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6399/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/4656/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/785/

File: docs/configuration-parameters.md [hunk 19]
```diff
@@ -58,6 +58,7 @@ This section provides the details of all the configurations required for the Car
...
 | spark.carbon.hive.schema.store | false | Carbondata currently supports 2 different types of metastores for storing schemas. This property specifies if Hive metastore is to be used for storing and retrieving table schemas                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+| carbon.enable.partitiondata.trash | false | This property is set to enable the feature of moving the deleted partition data to trash when dropped partition command is given
```
Comment (Indhumathi27) [inline docs/configuration-parameters.md:61]:
```suggestion
| carbon.enable.partitiondata.trash | false | This property, when enabled, will move the dropped partition data to trash on ALTER DROP PARTITION operation.
```

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 20]
```diff
@@ -2886,5 +2886,15 @@ private CarbonCommonConstants() {
 
   public static final String CARBON_ENABLE_MULTI_VERSION_TABLE_STATUS_DEFAULT = "false";
 
+  /**
+   * This flag is set to enable the feature of moving the partition data to trash
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:2890]:
```suggestion
   * Enable this property to move the dropped partition data to trash on ALTER DROP PARTITION operation
```

File: docs/ddl-of-carbondata.md [hunk 21]
```diff
@@ -1107,6 +1107,12 @@ Users can specify which columns to include and exclude for local dictionary gene
   ALTER TABLE locationTable DROP PARTITION (country = 'US');
   ```
 
+   The below flag is set to enable the feature of moving the deleted partition data to trash when dropped partition command is given.
```
Comment (Indhumathi27) [inline docs/ddl-of-carbondata.md:1110]: better to add it as a note
```suggestion
   **NOTE:** Enable [carbon.enable.partitiondata.trash](./configuration-parameters.md#system-configuration) to move dropped partition data to trash during alter table DROP PARTITION.
```

File: docs/faq.md [hunk 22]
```diff
@@ -481,4 +482,8 @@ TableStatusRecovery.main(args) --> args is of length two: 1. Database Name 2. Ta
 TableStatus Recovery tool cannot recover table status version files for the below two scenarios
 1. After compaction, if table status file is lost, cannot recover compacted commit transaction, as the lost version file only has merged load details.
 2. After Delete segment by Id/Date, if table status file is lost, cannot recover deleted segment commit transaction, as the lost version file only has the segment status as deleted.
-3. Table status recovery on materialized view table is not supported.
\ No newline at end of file
+3. Table status recovery on materialized view table is not supported.
+
+## Why deleted partition data still showing in file system
+Please enable the feature **moving deleted partition data to trash** by making the `carbon.enable.partitiondata.trash` property true.
```
Comment (Indhumathi27) [inline docs/faq.md:488]: Can mention like below.
By default, the dropped partition data will not be physically removed from the table store until the table is dropped. Enable [carbon.enable.partitiondata.trash](./configuration-parameters.md#system-configuration) property to move all the dropped partitions data to trash during alter table DROP PARTITION operation itself.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/6400/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 3.1, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_3.1/786/

**Comment by Indhumathi27** [comment]
LGTM
